packages/amplify-graphql-function-transformer/src/graphql-function-transformer.ts

```diff
@@ -195,3 +255,256 @@ const lambdaArnKey = (name: string, region?: string, accountId?: string): string
+ * This should come from `ctx.inputDocument` where ctx is {@link TransformerContextProvider}
+ */
+const validateSchemaDefinedEventInvocationResponseShape = (inputDocument: DocumentNode): void => {
+  // validate shape { success: Boolean! }
+  // We've already validated that the response type of the query / mutation
+  const responseTypeDefinitionNode = inputDocument.definitions.find(
+    (definitionNode) => isObjectTypeDefinitionNode(definitionNode) && definitionNode.name.value === eventInvocationResponse.typeName,
+  ) as ObjectTypeDefinitionNode | undefined;
+
+  if (!responseTypeDefinitionNode) {
+    // This implies an invalid GraphQL schema due to a defined return type ('EventInvocationResponse')
+    // being undefined in the Model Schema -- upstream validation should have already thrown a
+    // 'Schema validation failed. Unkown type "EventInvocationResponse"' error.
+    // We're doing this check here because upstream conditions can change in the future.
+    const errorMessage =
+      `Missing '${eventInvocationResponse.typeName}' definition. Add this type definition to your schema:\n` +
+      eventInvocationResponse.shapeDescription;
+    throw new InvalidDirectiveError(errorMessage);
+  }
+
+  const errorMessage =
+    `Invalid '${eventInvocationResponse.typeName}' definition. Update the type definition in your schema to:\n` +
+    `${eventInvocationResponse.shapeDescription}`;
+
+  // 'EventInvocationResponse' should contain only one field -- success: Boolean!
+  // We're doing this to:
+  // 1. safely access the first element below.
+  // 2. throw if > 1 fields are defined on the type.
+  const containsOneField = responseTypeDefinitionNode.fields?.length === 1;
+  if (!containsOneField) {
+    throw new InvalidDirectiveError(errorMessage);
+  }
+
+  const [expectedField] = responseTypeDefinitionNode.fields;
+
+  // At this point, we know there's only one field on the 'EventInvocationResponse' defined in the schema.
+  // Now we validate that this field has the shape 'success: Boolean!'
+  const schemaDefinedTypeHasValidShape =
+    expectedField.name.value === 'success' &&
+    expectedField.type.kind === Kind.NON_NULL_TYPE &&
+    expectedField.type.type.kind === Kind.NAMED_TYPE &&
+    expectedField.type.type.name.value === 'Boolean';
+
+  if (!schemaDefinedTypeHasValidShape) {
+    throw new InvalidDirectiveError(errorMessage);
+  }
+};
+
+/**
+ * Validates that the parent type is valid for 'invocationType: Event'
+ * - parent type === 'Query' | 'Mutation'
+ * This is done to ensure we don't accidentally add Event invocation type support to other types, e.g. `@model` or subscriptions.
+ * @param config the {@link FunctionDirectiveConfiguration} for 'invocationType: Event' definitions.
+ */
+const validateIsSupportedEventInvocationParentType = (config: FunctionDirectiveConfiguration): void => {
+  const isValidParentType = config.resolverTypeName === 'Query' || config.resolverTypeName === 'Mutation';
```

Comment:
nit: graphql-utils.ts has `isMutationType` and `isQueryType` utils rather than propagating string comparisons. Disclaimer: the util uses string comparisons.

Reply:
These utilities actually don't work here. 🤷 
"don't work" meaning:
```ts
config.resolverTypeName = 'Query';
isQueryType(config.resolverTypeName) // false
```